File: Chapter_1/transform/transform.java

/*
ID: 100021881
LANG: JAVA
PROG: transform
*/

import java.io.*;
import java.util.*;
public class transform {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new FileReader("transform.in"));
        PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter("transform.out")));
        int N = Integer.parseInt(br.readLine());
        char[][] before = new char[N][N];
        char[][] after = new char[N][N];
        
        for (int ii = 0; ii < N; ii++) {
            before[ii]=br.readLine().toCharArray();
        }
        for (int ii = 0; ii < N; ii++) {
            after[ii]=br.readLine().toCharArray();
        }

        int transformation = 0;
        if (equal(rotate(before), after)) {
            transformation=1;
        } else if (equal(rotate(rotate(before)), after)) {
            transformation=2;
        } else if (equal(rotate(rotate(rotate(before))), after)) {
            transformation=3;
        } else if (equal(reflect(before), after)) {
            transformation=4;
        } else if (equal(rotate(reflect(before)), after)) {
            transformation=5;
        } else if (equal(rotate(rotate(reflect(before))), after)) {
            transformation=5;
        } else if (equal(rotate(rotate(rotate(reflect(before)))), after)) {
            transformation=5;
        } else if (equal(before, after)) {
            transformation=6;
        } else {
            transformation=7;
        }
        out.println(transformation);
        out.close();
        br.close();
    }
    public static char[][] reflect(char[][] m) {
        int N = m.length;
        char[][] mat = new char[N][N];
        for (int ii = 0; ii < N; ii++) {
            for (int jj = 0; jj < N; jj++) {
                char ma = m[ii][N-jj-1];
                mat[ii][jj]=ma;
            }
        }
        return mat;
    }
    public static char[][] rotate(char[][] m) {
        int N = m.length;
        char[][] mat = new char[N][N];
        for (int x = 0; x < N / 2; x++) { 
            for (int y = x; y < N-x-1; y++) { 
                char before = m[x][y];
                mat[x][y] = m[y][N-1-x]; 
                mat[y][N-1-x] = m[N-1-x][N-1-y]; 
                mat[N-1-x][N-1-y] = m[N-1-y][x]; 
                mat[N-1-y][x] = before; 
            } 
        }
        if (N%2==1) {
            mat[(N-1)/2][(N-1)/2] = m[(N-1)/2][(N-1)/2];
        }
        return mat; 
    }
    public static boolean equal(char[][] a, char[][] b) {
        boolean output = true;
        for (int i = 0; i < a.length; i++) {
            for (int j = 0; j < a[0].length; j++) {
                if (a[i][j]!=b[i][j]) output = false;
            }
        }
        return output;
    }
}

File: Chapter_1/transform/transform.cpp
/*
ID: 100021881
LANG: C++
PROG: transform
*/


#include <iostream>
#include <fstream>
#include <cstring>

using namespace std;

int n = 10;
char before[10][10], after[10][10], temp[10][10];

void rotate() {
    for(int r = 0; r < n; r++) {
        for(int c = 0; c < n; c++) {
            after[c][n - 1 - r] = temp[r][c];
        }
    }
}

void reflect() {
    for(int r = 0; r < n; r++) {
        for(int c = 0; c < n; c++) {
            after[r][n - 1 - c] = temp[r][c];
        }
    }
}

int eqboard() {
    for(int r = 0; r < n; r++) {
        for(int c = 0; c < n; c++) {
            if(before[r][c] != after[r][c]) {
                return 0;
            }
        }
    }
    return 1;
}

int main() {
    ofstream fout("transform.out");
    ifstream fin("transform.in");
    int change = 7;
    fin >> n;
    for(int r = 0; r < n; r++) {
        for(int c = 0; c < n; c++) {
            fin >> before[r][c];
        }
    }
    for(int r = 0; r < n; r++) {
        for(int c = 0; c < n; c++) {
            fin >> after[r][c];
        }
    }
    memcpy(temp, before, sizeof(before));
    for(int i = 1; i <= 3; i++) {
        rotate();
        if(eqboard()) {
            change=i;
            break;
        }
    }
    memcpy(temp, before, sizeof(before));
    reflect();
    for(int i = 0; i <= 3; i++) {
        if(eqboard()) {
            change=4+i>0;
            break;
        }
        rotate();
    }
    memcpy(temp, before, sizeof(before));
    if (eqboard()) {
        change=6;
    }
    fout << change << endl;
    exit(0);
}
